zcash_client_backend/src/data_api/mem_wallet.rs

```diff
@@ -379,12 +381,62 @@ impl WalletWrite for MemoryWalletDb {
         todo!()
     }
 
+    /// Adds a sequence of blocks to the data store.
+    ///
+    /// Assumes blocks will be here in order.
     fn put_blocks(
         &mut self,
+        // TODO: Figure out what to do with this field.
         _from_state: &super::chain::ChainState,
-        _blocks: Vec<ScannedBlock<Self::AccountId>>,
+        blocks: Vec<ScannedBlock<Self::AccountId>>,
     ) -> Result<(), Self::Error> {
-        todo!()
+        // TODO:
+        // - Make sure blocks are coming in order.
+        // - Make sure the first block in the sequence is tip + 1?
+        // - Add a check to make sure the blocks are not already in the data store.
+        for block in blocks.into_iter() {
+            let mut transactions = HashMap::new();
+            for transaction in block.transactions().into_iter().cloned() {
+                let txid = transaction.txid();
+                let account_id = 0; // TODO: Assuming the account is 0, handle this accordingly.
+                let ufvk = self
+                    .accounts
+                    .get(&account_id)
+                    .ok_or(Error::AccountUnknown(0))?
+                    .ufvk
+                    .sapling()
+                    .ok_or(Error::ViewingKeyNotFound(0))?;
+                let nk = ufvk.to_nk(Scope::External);
+
+                let spent_nullifiers = transaction
+                    .sapling_outputs()
+                    .iter()
+                    .map(|o| {
+                        let nullifier = o.note().nf(&nk, o.note_commitment_tree_position().into());
+                        // TODO: Populate the bool field properly.
+                        self.sapling_spends.entry(nullifier).or_insert((txid, true));
+                        nullifier
+                    })
+                    .count();
+
+                // Is `self.tx_idx` field filled with all the transaction ids from the scanned blocks ?
+                self.tx_idx.insert(txid, block.block_height);
+                transactions.insert(txid, transaction);
+            }
+
+            let memory_block = MemoryWalletBlock {
+                height: block.block_height,
+                hash: block.block_hash,
+                block_time: block.block_time,
+                transactions,
+                // TODO: Add memos
+                memos: HashMap::new(),
+            };
+
```

Comment:
This will have several additional responsibilities related to what will need to happen in the wallet overall:
* Note commitment trees will need to be updated.
* Received notes need to be made available for note selection & balance calculation
* Spent notes need to be made unavailable for note selection & balance calculation

Reply:
With respect to adding to the note commitment tree, it looks like you're assuming linear scanning here. That's probably okay for a v0, because it saves a bunch of complexity; I just haven't been thinking about that use case in a long time; if that's the case though, you won't be able to use the sync engine from https://github.com/zcash/librustzcash/pull/1217 

With respect to the other two comments:
Depending upon how balance is calculated, it may be that only the second comment is relevant; when a note is spent, we need to update our record for that note so that it's no longer included as having value belonging to the wallet.